### ISSUE TITLE
torch.compile() drops transformer/Qwen1.5-7B model output quality from good to unusable

### ISSUE THREAD

**Issue by Qubitium** [pytorch/pytorch#120462]
### 🐛 Describe the bug

Not sure how we can further debug this and provide more data but we are seeing drastic output quality difference between no-torch.compile and torch.compile.  We are new to torch.compile().

Does torch.compile inherently cause accuracy issues? Not sure if just natural part of the beast. 

model load code:
```
model = AutoModelForCausalLM.from_pretrained(
    adapter_path,
    device_map="cuda",
    config=config,
    torch_dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",

# if we comment the following line, output quality returns to normal
model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
```

gen code and gen-params are same for generation. 

The difference is as if we are running a vastly different, a much more inaccurate model when torch.compile() is enabled. The difference, drop in quality, is not related to randomness of sampling.  

### Versions

``` 
model: finetuned Qwen1.5-7B
transformers: 3.38.1
pytorch: 2.2
accelerate: 0.27.2
flash-attn: 2.5.3
os: ubuntu 22.04
gpu: 4090
```

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519

**Comment by ezyang** [comment]
I'm putting this up for triage review because there is an interesting question here, which is that torch.compile numerics are *necessarily* different from eager (we do not guarantee they are exactly the same), but what I am not sure is if we consider it a bug if you fine tune a model with eager pytorch, and then torch.compile'd version is unusable, is this a bug in torch.compile or could it be that this is expected and you needed to fine tune the model with torch.compile?

That being said, since this is a transformer accuracy thing, it's possible that https://github.com/pytorch/pytorch/pull/119500 fixes the problem. I'm not sure if this one made it to 2.2, you can try a nightly. @bdhirsh 

We might be willing to take a closer look; more detailed repro instructions will help in that case.

**Comment by Chillee** [comment]
We discussed this in triage review, and we do not believe that (modulo bugs), `torch.compile` should ever result in *worse* accuracy, particularly inference.

**Comment by siddhatiwari** [comment]
I've also gotten much worse accuracy using torch.compile, here's a repro (for a Text-to-Speech model). 

With the following changes, the final speech output is much worse. Same worse results when using other torch.compile parameters like `mode="max-autotune"`, `mode="max-autotune-no-cudagraphs"`, `fullgraph=True`, `fullgraph=False`, `dynamic=True`, `dynamic=False`, and also when torch.compile is applied to other modules of the Decoder class.

https://github.com/siddhatiwari/StyleTTS2/commit/13a9f87e97af98561566b643728e01608acc1bb1

**Comment by ezyang** [comment]
Thanks @siddhatiwari, any chance you can give instructions for how to reproduce your results on your repo?

**Comment by siddhatiwari** [comment]
StyleTTS2 is a model for generating speech from text. It doesn't currently use torch.compile in any capacity. I tried modifying it by adding a single torch.compile to a small part of the overall model. You can find the original repo https://github.com/yl4579/StyleTTS2 and my modified repo https://github.com/siddhatiwari/styleTTS2. 

Note that you only need to clone my modified repo to generate audio with both the unmodified model (original quality) and the torch.compile model (with lower quality).

The following steps explain how to set everything up, produce good outputs with the unmodified model, and bad outputs with the torch.compile model:

Part 1: Setup
-----------------------------------------------------------
Run each command:
```
git clone https://github.com/siddhatiwari/styleTTS2
cd styleTTS2
python3 -m venv venv && source venv/bin/activate
pip install SoundFile torchaudio munch torch pydub pyyaml librosa nltk matplotlib accelerate transformers phonemizer einops einops-exts tqdm typing-extensions git+https://github.com/resemble-ai/monotonic_align.git
sudo apt-get install espeak-ng
sudo apt-get install git-lfs
git-lfs clone https://huggingface.co/yl4579/StyleTTS2-LibriTTS
mv StyleTTS2-LibriTTS/Models .
mv StyleTTS2-LibriTTS/reference_audio.zip .
unzip reference_audio.zip
mv reference_audio Demo/reference_audio
```


Part 2: Generating audio using unmodified model
-----------------------------------------------------------
Run:
```
python3 repro.py 
```
See generated audio:
1789_142896.wav
696_92939.wav


Part 3: Generating audio using torch.compile model
-----------------------------------------------------------
In ./Modules/hifigan.py (https://github.com/siddhatiwari/StyleTTS2/blob/main/Modules/hifigan.py#L436):

Comment line:
```
# self.N_conv = weight_norm(nn.Conv1d(1, 1, kernel_size=3, stride=2, groups=1, padding=1))
```
Uncomment lines:
```
self.N_conv = torch.compile(weight_norm(
    nn.Conv1d(1, 1, kernel_size=3, stride=2, groups=1, padding=1)
), mode="reduce-overhead", fullgraph=True, dynamic=False)
```

Run:
```
python3 repro.py 
```
See generated audio:
1789_142896.wav
696_92939.wav

You will notice that the generated audio in Part 3 is worse quality than that in Part 2.

-----------------------------------------------------------


Sample audio outputs you should get after following these steps:
https://github.com/siddhatiwari/StyleTTS2/tree/main/samples

**Comment by siddhatiwari** [comment]
@ezyang Hey, just wanted to follow up on this. Any ideas on what I could do on my end to help you guys investigate this further?

**Comment by ezyang** [comment]
Hi @siddhatiwari, some basic troubleshooting ala https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html would be helpful. What I find most useful is to do an ablation: does it happen if you get rid of reduce-overhead, what about backend="aot_eager" or backend="eager"?
